src/providers/ows/qgsgeocmsdataitems.cpp

```diff
@@ -0,0 +1,267 @@
+ ***************************************************************************/
+
+#include "qgsproviderregistry.h"
+#include "qgsowsdataitems.h"
+#include "qgslogger.h"
+#include "qgsnewhttpconnection.h"
+#include "qgsgeonodenewconnection.h"
+#include "qgsgeocmsdataitems.h"
+
+
+typedef QList<QgsDataItemProvider *> dataItemProviders_t();
+
+QgsGeoCMSConnectionItem::QgsGeoCMSConnectionItem( QgsDataItem *parent, QString name, QString path, QgsGeoCMSConnection *conn )
+  : QgsDataCollectionItem( parent, name, path )
+  , mGeoCMSName( parent->name() )
+  , mUri( conn->uri().uri() )
+  , mConnection( conn )
+{
+  mIconName = QStringLiteral( "mIconConnect.png" );
+}
+
+QVector<QgsDataItem *> QgsGeoCMSConnectionItem::createChildren()
+{
+  QVector<QgsDataItem *> services;
+
+  QStringList wmsUrl = mConnection->serviceUrl( QStringLiteral( "WMS" ) );
+  QStringList wfsUrl = mConnection->serviceUrl( QStringLiteral( "WFS" ) );
+  QStringList xyzUrl = mConnection->serviceUrl( QStringLiteral( "XYZ" ) );
+
+  if ( !wmsUrl.isEmpty() )
+  {
+    QString path = mPath + "/wms";
+    QgsDataItem *service = new QgsGeoCMSServiceItem( this, mConnection, QString( "WMS" ), path );
+    services.append( service );
+  }
+
+  if ( !wfsUrl.isEmpty() )
+  {
+    QString path = mPath + "/wfs";
+    QgsDataItem *service = new QgsGeoCMSServiceItem( this, mConnection, QString( "WFS" ), path );
+    services.append( service );
+  }
+
+  if ( !xyzUrl.isEmpty() )
+  {
+    QString path = mPath + "/xyz";
+    QgsDataItem *service = new QgsGeoCMSServiceItem( this, mConnection, QString( "XYZ" ), path );
+    services.append( service );
+  }
+
+  return services;
+}
+
+QList<QAction *> QgsGeoCMSConnectionItem::actions()
+{
+  QAction *actionEdit = new QAction( tr( "Edit..." ), this );
+  QAction *actionDelete = new QAction( tr( "Delete" ), this );
+  connect( actionEdit, &QAction::triggered, this, &QgsGeoCMSConnectionItem::editConnection );
+  connect( actionDelete, &QAction::triggered, this, &QgsGeoCMSConnectionItem::deleteConnection );
+  return QList<QAction *>() << actionEdit << actionDelete;
+}
+
+void QgsGeoCMSConnectionItem::editConnection()
+{
+  QgsGeoNodeNewConnection *nc = new QgsGeoNodeNewConnection( nullptr, mConnection->mConnName );
+  nc->setWindowTitle( tr( "Modify GeoCMS connection" ) );
+
+  if ( nc->exec() )
+  {
+    // the parent should be updated
+    mParent->refresh();
+  }
+}
+
+QgsGeoCMSServiceItem::QgsGeoCMSServiceItem( QgsDataItem *parent, QgsGeoCMSConnection *conn, QString serviceName, QString path )
+  : QgsDataCollectionItem( parent, serviceName, path )
+  , mName( conn->mConnName )
+  , mServiceName( serviceName )
+  , mConnection( conn )
+{
+  if ( serviceName == "WMS" || serviceName == "XYZ" )
+  {
+    mIconName = QStringLiteral( "mIconWms.svg" );
+  }
+  else
+  {
+    mIconName = QStringLiteral( "mIconWfs.svg" );
+  }
+}
+
+QVector<QgsDataItem *> QgsGeoCMSServiceItem::createChildren()
```

Comment:
There are a lot of debug messages printed in loops within this method. Is this very noisy in real life? If yes, can you remove it or move it to a different message log level?